chia/simulator/full_node_simulator.py

```diff
@@ -144,3 +203,288 @@ async def reorg_from_index_to_new_index(self, request: ReorgProtocol):
+        if height_before is None:
+            height_before = uint32(0)
+
+        for count in itertools.count(1):
+            height = uint32(height_before + count)
+            rewards += calculate_pool_reward(height) + calculate_base_farmer_reward(height)
+
+            if rewards >= amount:
+                break
+        else:
+            raise Exception("internal error")
+
+        if isinstance(timeout, _Default):
+            timeout = (count + 1) * timeout_per_block
+
+        with anyio.fail_after(delay=timeout):
+            await self.farm_blocks_to_wallet(count=count, wallet=wallet, timeout=None)
+            return rewards
+
+    async def wait_transaction_records_entered_mempool(
+        self,
+        records: Collection[TransactionRecord],
+        timeout: Union[None, float] = 5,
+    ) -> None:
+        """Wait until the transaction records have entered the mempool.  Transaction
+        records with no spend bundle are ignored.
+
+        Arguments:
+            records: The transaction records to wait for.
+        """
+        with anyio.fail_after(delay=timeout):
+            ids_to_check: Set[bytes32] = set()
+            for record in records:
+                if record.spend_bundle is None:
+                    continue
+
+                ids_to_check.add(record.spend_bundle.name())
+
+            for backoff in backoff_times():
+                found = set()
+                for spend_bundle_name in ids_to_check:
+                    tx = self.full_node.mempool_manager.get_spendbundle(spend_bundle_name)
+                    if tx is not None:
+                        found.add(spend_bundle_name)
+                ids_to_check = ids_to_check.difference(found)
+
+                if len(ids_to_check) == 0:
+                    return
+
+                await asyncio.sleep(backoff)
+
+    async def process_transaction_records(
+        self,
+        records: Collection[TransactionRecord],
+        timeout: Union[None, float] = (2 * timeout_per_block) + 5,
+    ) -> None:
+        """Process the specified transaction records and wait until they have been
+        included in a block.
+
+        Arguments:
+            records: The transaction records to process.
+        """
+        with anyio.fail_after(delay=timeout):
+            coins_to_wait_for: Set[Coin] = set()
+            for record in records:
+                if record.spend_bundle is None:
+                    continue
+
+                coins_to_wait_for.update(record.spend_bundle.additions())
+
+            coin_store = self.full_node.coin_store
+
+            await self.wait_transaction_records_entered_mempool(records=records, timeout=None)
+
+            while True:
+                await self.farm_blocks_to_puzzlehash(count=1, timeout=None)
+
+                found: Set[Coin] = set()
+                for coin in coins_to_wait_for:
+                    # TODO: is this the proper check?
```

Comment:
check for removals and additions?

Reply:
Pre-existing `TODO:`, can deal with separately at some point.